sm-c.man
.\" Man pages for sm-c (sendmail-connector)

.TH man 1 "2014-03-07" "1.5" "sm-c (sendmail-connector) man pages"

.SH NAME
.B "sm-c (sendmail-connector)"

The "sm-c" package provides a sendmail-connector script (sm-c) that comes between the sendmail symbolic link and the msmtp, an SMTP client. Effectively, sm-c connects a mail user agent (MUA) such as s-nails' or heirloom-mailx's mail(x) to msmtp.

.SH SYNOPSIS
Invoked automatically by any program calling sendmail as through the installation process, the sendmail symbolic link points at sm-c. For example, the mail(x) program provided by s-nail or heirloom-mailx typically calls sm-c with the following parameters:
.br
.RS
.B sm-c
-i \fB<recipient 1>
\fR[\fB<recipient 2>\fR] [\fB...\fR] [\fB<recipient N>\fR]
.RE
.PP
where the recipients are users' login names or e-mail address.
.PP
\fBsm-c\fR can be invoked manually the way msmtp can.
.br

.SH "DESCRIPTION"
Sendmail-connector
.IP \[bu]
replaces users' login names with their e-mail addresses as specified in /etc/aliases;
.br
If a login name is not defined in /etc/aliases sm-c drops such a recipient.
.IP \[bu]
gives debugging data showing the e-mail message source and recipients;
.br
The debugging is switched on with an option in /etc/sm-c.conf or ~/.sm-c.conf.
.br
   debug=1 (or debug=yes)
.br
The file is not created in the installation process. It has to be created by user if debugging is supposed to work.
.br
The debugging log is made into /tmp/sm-${USER}-$(date +%H%M%S).log files.
.IP \[bu]
comes in handy when cron(ie) reports events or notifies about errors;
.IP \[bu]
corrects the charset in the contents of e-mail message;
.br
UTF-8 charset is hard coded into the script. Find the "CS" variable and change it to your liking.
.br
The language/charset settings for cron/cronie are different from those set in the linux system. Put these settings in your user cronjobs file (crontab -e).
.br
Here is an example with Polish language settings.
.br

   LANG=pl_PL.UTF-8
.br
   LANGUAGE=pl
.br
   LC_CTYPE=pl_PL.UTF-8
.br

If you do not put these into your cronjobs file, you will not be able to send mail if there are any "language" characters in the subject line. This script will help you avoid problems with charset in the contents of your mail even if you do not set your language for cron, but will not help if special characters are in your subject line.



.SH OPTIONS AND PARAMETERS

The options used with sm-c are those used with msmtp.
.br
There is only one option extra/specific for sm-c: the '-Z' option.
.TP
.B -Z
.br
This option adds a read-confirmation request to the e-mail message header.
.br
\fBThe "-Z" flag can only be used when it is passed to sm-c by an MUA like mail(x).\fR Other MUA's were not tested. \fBA direct manual passing -Z to sm-c does not work.\fR This would be an extra msmtp feature, but it has not been implemented. The disposition to request for a read confirmation is added to the e-mail header created by MUA (e.g. mail(x)) instead.
.RS

An example of passing -Z to sm-c:

.br
	mailx -O -Z
.br
	mailx -O-Z
.br

\fBNote that this is different from DSN (delivery status notification)\fR. A DSN is requested by the "-N" msmtp option. See the mail(x) "-O" option below and read the msmtp man pages.
.RE
.PP
Other options:
.TP
.B -i
.br
This option is the first option that \fUmail(x)\fR passes to sm-c (and then to msmtp).
.TP
.B -FCronDaemon
If sm-c is invoked by \fUcron(ie)\fR notification mechanism, "\fB-FCronDaemon\fR" is the first option passed to sm-c/msmtp.
.TP
.B recipient
.br
The recipient should be a proper (domain) e-mail address. Any host system $USER may also be used providing the \fI/etc/aliases\fR file holds the \fIproper address(es)\fR for the $USER. If there are no e-mail addresses for the $USER defined in /etc/aliases, such $USER is dropped from the recipients list.

.PP
When sm-c is called manually, try using options as defined for msmtp.

.TP
.B -t
.br
The "-t" option tells msmtp to source TO, CC, BCC addresses from the message contents passed. The "-t" option works the same way with mail(x).

.PP
It is possible to pass extra options from mail(x) to sm-c by using the "-O" option. However, it is important that if the option that is being passed has additional attributes those attributes must be "glued to the msmtp's -O option, e.g.
.br

      mail -t \fB-O -Nsuccess,failure,delay\fR

.br
will pass "-Nsuccess,failure,delay" to sm-c/msmtp. It requests a DSN (delivery-state-notification). None of servers used by AUTHOR had the DSN feature enabled, though.

.PP
.SH DEPENDENCIES AND PROGRAMS SM-C WORKS WITH
.PP
Dependencies include typical functions:  grep, sed.
.PP
As sm-c (sendmail-connector) is a connector, it needs programs it connects. Whether they are technically dependencies is a matter of semantics. The author thinks them dependencies.
.TP
.B msmtp
sm-c requires the msmtp package to be installed. Other smtp clients were not tested, but may work except for the -Z option described above.
.TP
\fBs-nail\fR/\fBheirloom-mailx\fR providing mail/mailx programs [optional, highly recommended]
.br
In order to be able to send e-mails with attachments and to request read conformations (the -Z flag), the user has to install the s-nail package (or heirloom-mailx).

.SH INSTALLATION
As sm-c is a bash script it could be installed manually by creating a sendmail symbolic link (/usr/bin/sendmail) or replacing an existing one with a new one that points at sm-c.
.PP
The automatic installation with PKGBUILD checks conflicts and \fBinstalls msmtp as a dependency\fR (!). It also installs this manual.
.br
Note that the s-nail/heirloom-mailx package is not installed automatically.

.SH CONFIGURATION, USE AND EXAMPLES
.PP
.IP \[bu]
Configuration file \fI~/.config/sm-c/sm-c.conf\fR (alternatively \fI/etc/sm-c.conf\fR) holds only one option:
.RS
debug=1 (or yes/Yes; anything else means 'no')
.RE
.IP \[bu]
There is no need to make cron/cronie call sm-c (or msmtp) directly. Do not change anything and just let cron/cronie call sendmail, which it does by default.
.IP \[bu]
The same regards mail(x) - do not make it call sm-c or msmtp. Let is call sendmail, and modify only the sendmail link if necessary. The installation process does it for you.
.IP \[bu]
For sm-c to work correctly with cron/cronie, you must put your language settings into your cronjobs file. Run "crontab -e" and put these or similar settings at the beginning of the file. These settings are for the Polish language.
.RS
      LANG=pl_PL.UTF-8
.br
      LANGUAGE=pl
.br
      LC_CTYPE=pl_PL.UTF-8
.br
Failing to do that may result in your not being able to send e-mail when the subject line has your national language special characters. It is caused by the system, not sm-c.
.RE
.IP \[bu]
Note that when used by cron/cronie, sm-c will be sending to the $USER all output of commands in the cronjob file. To avoid it use '>/dev/null' after each command in the cronjobs file (crontab).
.IP \[bu]
If a login/user name is not defined in /etc/aliases, sm-c drops such a recipient.

.SH TYPICAL CONTENTS OF .msmtprc (EXAMPLE)
# Accounts will inherit settings from this section
.br
defaults
.br
logfile /tmp/msmtp.log
.br
auth           on
.br
tls            on
.br
#aliases       /etc/aliases 		# This line should rather be in /etc/msmtprc
.br

.br
\fBaccount        gmail\fR
.br
host           smtp.gmail.com
.br
port           587
.br
user           user@gmail.com
.br
passwordeval   "gpg -d /path/to/password_file1"
.br
from           user@gmail.com
.br
tls_trust_file /usr/share/ca-certificates/mozilla/Equifax_Secure_CA.crt
.br

# Onet email service
.br
\fBaccount 		onet\fR
.br
host 		smtp.poczta.onet.pl
.br
port 		587
.br
user 		user2@poczta.onet.pl
.br
password   	"your_secret_password"
.br
from  		user2@poczta.onet.pl
.br
tls_trust_file /etc/ssl/certs/ca-certificates.crt
.br

\fBaccount 	 	other\fR
.br
host       	mail.domain.com
.br
port       	587
.br
user       	other_user
.br
passwordeval   "gpg -d /path/to/password_file2"
.br
from       	other_user@domain.com
.br
tls_fingerprint AA:BB:CC:DD:EE:FF:00:11:22:33:44:55:66:77:88:99:AA:BB:CC:DD
.br
# See \fIwww.madboa.com/geek/openssl/#cert-retrieve\fR
.br
# for info on retrieving certificates and fingerprints.
.br

# Set a default account
.br
\fBaccount default\fR : other
.br

.SH TYPICAL CONTENTS OF .mailrc (EXAMPLE THAT GOES WITH THE EXAMPLE ABOVE)
.br
# $ mail -s "subject line" -a /path/file recipient@email.com < /path/body.txt
.br
# cat /path/body.txt | $ mail -s "subject line" -a /path/file recipient@email.com
.br
# cat /path/header+body.txt | $ mail -t
.br

\fBaccount default\fR {
.br
set from="name1@domain.org"
.br
set message-sendmail-extra-arguments="-a other"
.br
}
.br

\fBaccount gmail\fR {
.br
set from="name2@gmail.com"
.br
set message-sendmail-extra-arguments="-a gmail"
.br
}
.br

\fBaccount onet\fR {
.br
set from="name3@poczta.onet.pl"
.br
set sendmail=/usr/bin/msmtp   				# Don't do it.
.br
set message-sendmail-extra-arguments="-a onet -S ssl-verify=ignore"
.br
}
.br

\fBaccount other\fR {
.br
set from="name4@other.org"
.br
#set ttycharset=utf-8
.br
#set sendcharsets=utf-8
.br
#set encoding=8bit
.br
set message-sendmail-extra-arguments="-a other"
.br
}
.br

.SH HISTORY
.PP
The idea to create sm-c was based on Jim Lofft's script find_alias_for_msmtp.sh (06/19/2009) and the changes to the script by Ovidiu Constantin <ovidiu@mybox.ro> && http://blog.mybox.ro.
.PP
The sm-c script was first a heavily modified version of find_alias_for_mstp.sh (July 2013). Tests were made with mailx and cron/cronie (both calling /usr/bin/sendmail, a symlink to msmtp).
In March 2014 the -Z option was added and further modification made to effectively turn sm-c into a new script.

.SH SEE ALSO
.IP \[bu]
Read about the sendmail's aliases file (/etc/aliases). For instance here:
.RS
.I 	http://www.gsp.com/cgi-bin/man.cgi?topic=aliases
.RE
.IP \[bu]
msmtp man pages and the files in /usr/share/doc/msmtp
.IP \[bu]
s-nail (heirloom-mailx) man pages
.IP \[bu]
Read
.I http://www.madboa.com/geek/openssl
about retrieving certificates and fingerprints.

.SH BUGS
Please contact the author if you find any.

.SH AUTHOR
.AU
Wiesław Magusiak <w.magusiak at gmail . com>

.SH COPYRIGHT
No warranty is given, and no responsible is taken for the use.
.br
Latest GPL.


sm-c.sh
#!/bin/bash
# sm-c.sh v1.1 by Wiesław Magusiak (2013-07-31)
# sm-c.sh v1.5.1 (2014-03-08)
# sm-c.sh v1.5.2 (2014-03-14): ~/.sm-c.conf moved to ~/.config/sm-c/sm-c.conf
# A "sendmail" replacement. Requires "msmtp" installed.
# Put this script anywhere (e.g. /usr/local/bin/) and create a symbolic link:
#
# 	# ln -s /usr/local/bin/find_email_aliases.sh /usr/bin/sendmail
#
# If called from cron/cronie, the script scans the /etc/alias file for valid domain
# email addresses and calls msmtp (http://msmtp.sourceforge.net/).
#
# Some advice:
# (1) Do not make cron/cronie call msmtp instead of sendmail; Calling sendmail is default.
# (2) Do not make mailx call msmtp. By default it will call sendmail.
# (3) If you install this script with its package, it will check dependecies and conflicts,
#     and will create sendmail as a symbolic link to this script.
#     Otherwise, put the script somewhere (e.g. /usr/local/bin/) and create the symbolic
#     manually.
# (4) The language/charset settings for cron/cronie are different from those
#     set in the linux system. Put these settings in your user cronjobs file (crontab -e).
#     Here is an example with Polish language settings.
#       LANG=pl_PL.UTF-8
#       LANGUAGE=pl
#       LC_CTYPE=pl_PL.UTF-8
#     Failing to do that may lead to your not being able to send e-mail.
#     It is an error in system, not in sm-c.
# (5) With this script and mailx, you will be able to send e-mail messages with attachments
#     from cron/cronie.
# ---------------------------------------------------------------------------------- #
# This is a fully functional script stuffed with a lot of lines for debugging.       #
# Check "$DEBUG" after mailx has been called directly or cron sent an email message. #
# You can safely remove all "debugging parts" from the script below.                 #
# ---------------------------------------------------------------------------------- #

#----Create a log file for debugging------------
if [[ -f ~/.sm-c.conf ]]; then
	D=$(grep ^debug ~/.config/sm-c/sm-c.conf|cut -d= -f2)	# set debug={1, y, Y, yes,...}
elif [[ -f /etc/sm-c.conf ]]; then
	D=$(grep ^debug /etc/sm-c.conf|cut -d= -f2)		# set debug={1, y, Y, yes, Yes, yEs,...}
fi

[[ "$D" == "1" || "$D" == "[Yy]" || "$D" == "[yY][eE][sS]" ]] \
	&& DEBUG=/tmp/sm-${U}-$(date +%H%M%S).log || DEBUG="/dev/null"
#----Create a log file for debugging--END-------

PARAMS=" $@ "		# Parameters sendmail is called with
TO=0				# E-mail address(es)
MSG="$(cat)"		# Message contents passed to sendmail
CS="charset=utf-8"	# Charset for email.
					# Needs to be set when mailx is called by cron/cronie.
U=$USER 			# If sendmail is called from cron, U=""

# -Z - the only 'native' sm-c flag/option
if [[ "$PARAMS" = *-Z* ]]; then 		# if '-Z' flag passed, ask for a read receipt
	MSMTP_CONFF=~/.msmtprc  			# msmtp user configuration file path
	FROM=$(grep "^account default" $MSMTP_CONFF | sed 's/.*: //') 	# default msmtp account
	FROM=$(sed -n "/$FROM/,/^$/p" $MSMTP_CONFF | grep from) 	# line with 'from' addres
	FROM=$(echo $FROM |sed 's/.* //g') 							# the 'from' e-mail address
	MSG="Disposition-Notification-To: $FROM\n"${MSG}
	PARAMS=${PARAMS/-Z /} 				# remove '-Z' flag from parameters
fi

PREFIX=${PARAMS% -*}
if [[ "$PARAMS" = "$PREFIX" ]]; then
	PREFIX=""
	RECIP="$PARAMS"
else
	RECIP=${PARAMS#$PREFIX }
	RECIP=${RECIP#-* }
	PREFIX=${PARAMS% $RECIP}
fi 					# RECIP holds recipient names/addresses; PREFIX holds options
z=""
# Find e-mail addresses for recipient names in /etc/aliases
if [[ -f /etc/aliases ]]; then  	# Does the file exist?
	if [[ -n "$RECIP" ]]; then
		# Make sure recipient addresses are "@domain" style ones
		for x in $RECIP; do 		# Important! No quotation marks here!
			if [[ $x = *@* ]]; then
				z=$z" $x"
			else
				y="$(grep "$x:" /etc/aliases|cut -d: -f2|sed 's/\, / /g'|sed 's/ *$//g')"
				# Recepient without "@domain" who is not listed in /etc/aliases gets eliminated.
				z=$z"$y"
			fi
		done 						# z has a leading space
	fi
else
		for x in $RECIP; do 		# Important! No quotation marks here!
			if [[ $x = *@* ]]; then
				z=$z" $x"
			else
				# Recepient without "@domain" who is not listed in /etc/aliases gets eliminated.
				: 					# continue
			fi
		done 						# z has a leading space
fi
PARAMS="${PREFIX# }${z}" 			# PREFIX holds options; z holds addresses (may be empty, too)

SMTPF="$(which msmtp)"

# ---- Debugging part ------------------
if [ -n "$DEBUG" ]; then
	#echo -e "The sendmail call by $(ps -p $(ps -p $$ -o ppid=) -o comm=):\n\t$0 $@" > "$DEBUG"
	echo -e "$(LANG=C date)\nThe sendmail call:\n\t$0 $@" > "$DEBUG"
	echo -n "called from " >> "$DEBUG"
	[[ -t 1 ]] && echo -e "a terminal." >> "$DEBUG" || echo "not a terminal" >> "$DEBUG"
	#[[ -t 1 || -t 0 ]] && echo "a terminal" >> "$DEBUG" || echo "not a terminal" >> "$DEBUG"
	echo "-------------------" >> "$DEBUG"
	echo -e "USER=${U} \tLANG=$LANG" >> "$DEBUG"
	echo -e "PID  \$\$    = $$\t$(ps -p $$ -o comm=)" >> "$DEBUG"
	echo -e "PPID \$PPID = $PPID\t$(ps -p $PPID -o comm=)" >> "$DEBUG"
	#cat /proc/$$/status | grep PPid: | grep -o "[0-9]*" >> "$DEBUG"
	#cat /proc/$$/status | grep Pid: | grep -o "[0-9]*" >> "$DEBUG"
	echo "-------------------" >> "$DEBUG"
fi
# ---- End of debugging part -----------

# mailx calls sendmail with -i as the first parameter and addresses as the following parameters
# Warning: mailx sorts parameters moving those with hyphens before those without hyphens!
# Advice: If you pass parameters with -O, put no spaces between -O and the past paarameters/valuses.
# 
if [  "$1" = '-i'  ]; then 		# Strong assumption: Called by mailx
	# No shift of parameters to include the "-i" flag in $TO
	#shift
	TO="$PARAMS"
	# If mailx is called from cron and no language settings are put into user's cronjobs file,
    # the "Content-Type" part of an email header has to be changed
	# from  Content-Type: application/octet-stream
	#   to  Content-Type: text/plain; charset=utf-8
	# Without it, thunderbird will not show your message "inline".
	MSG="${MSG/Type: application\/octet-stream/Type: text\/plain; $CS}"
	# ---- Debugging part ------------------
	if [ -n "$DEBUG" ]; then
        # Addressee as specified in $MSG, in the "To: " line.
		kk="$(echo "$MSG" | grep -m 1 'To: ')"
        kk="${kk:4}"
		if [[ -n "$kk" ]]; then
			echo -e "Addressee as in the email-header lines:\n$kk"  >> "$DEBUG"
			echo "-------------------" >> "$DEBUG"
		fi
	fi
	# ---- End of debugging part -----------

# crone/cronie calls sendmail with -FConeDaemon as the first parameter
# 	(-FCronDaemon -i -odi -oem -oi -t -f <user>)
elif [ "$1" = '-FCronDaemon' ]; then
	RECIP="$(echo "${@:$#}")"
	# ---- Debugging part ------------------
	if [ -n "$DEBUG" ]; then
		echo "\$RECIP=$RECIP"  >> "$DEBUG"
	fi
	# ---- End of debugging part -----------
	if [ -f /etc/aliases ]; then
		#-----Find email addresses and remove commas between them if any.
		TO="$(grep "$RECIP:" /etc/aliases|cut -d: -f2|sed "s/\, / /g"|sed "s/ *$//g")"
		TO="-i$TO"	# $TO starts with a space here.
            #-----Unnecessary, but satisfies my aesthetic taste (in thunderbird)------------------------
			TO1="$(echo "$TO"|cut -d\  -f2)"        # To have just one sender/adressee displayed
			MSG="${MSG/To: $RECIP/To: $TO1}"		# Replace "To: $RECIP" with "To: $TO"
			MSG="${MSG/From: $RECIP/From: $TO1}"	# Replace "From: $RECIP" with "From: $TO"
            #-----The end of "Unnecessary or even wrong"------------------------------------------------
		#-----An option; a full parameters line
        # (It does not work quite well, sending email to one addressee from cron, not all of them.)
		# TO=${PARAMS/-f $RECIP/-f $TO}
	else
		echo "Missing file:  /etc/aliases."
		exit 1
		#-----An alternative solution----
		# DEFAULTEMAIL="a_VIP@domain.com"
		# TO="-i "$DEFAULTEMAIL
	fi
	# ---- Debugging part ------------------
		#echo \$RECIP=$RECIP  >> "$DEBUG"
	# ---- End of debugging part -----------

# other programs
elif [ "$1" = '-t' ]; then  	# This and the next "else" could be combined, but
	if [ -n "$DEBUG" ]; then 	# I want them separate for at least debugging reasons
		echo "Mail sent with the -t parameter" >> "$DEBUG"
		echo "-------------------" >> "$DEBUG"
	fi
	TO="$PARAMS" 		# -t draws addresses from $MSG automatically
else
	if [ -n "$DEBUG" ]; then
		echo "* Mail sent without parameters or not recognised ones." >> "$DEBUG"
		echo "* Most likely the message will be prepared but NOT SENT." >> "$DEBUG"
		echo "* Expect an error message like 'No recepients found'" >> "$DEBUG"
		echo "-------------------" >> "$DEBUG"
	fi
	TO="$PARAMS" 	# Just the original parameter(s) sendmail was called with.
fi

# ---- Debugging part ------------------
if [ -n "$DEBUG" ]; then
	echo "The arguments for ${SMTPF##*/}: $TO" >> "$DEBUG"
	echo -------e-mail_structure------- >> "$DEBUG"
	echo -e "$MSG" >> "$DEBUG"
	echo -------end_of_email----------- >> "$DEBUG"
fi
# ---- End of debugging part -----------

echo -e "$MSG" | "$SMTPF" $TO
